File: app/src/main/java/com/msaggik/algorithmsdatastructures/java/search/BinarySearch.java
package com.msaggik.algorithmsdatastructures.java.search;

/**
 * Бинарный поиск
 * применим к последовательности отсортированных значений
 * Вычислительная сложность: О(log(N))
 * <p>
 * Пример: поиск значения в массиве
 * (если имеется возвращает его индекс, если отсутствует возвращает -1)
 **/

public class BinarySearch {
    public static int isValueInArray(int[] arrayInt, int value) {
        int left = 0, right = arrayInt.length - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            int current = arrayInt[middle];
            if (current == value) return middle;
            else if (current < value) left = middle + 1;
            else right = middle - 1;
        }
        return -1;
    }
}

File: app/src/main/java/com/msaggik/algorithmsdatastructures/kotlin/search/BinarySearch.kt
package com.msaggik.algorithmsdatastructures.kotlin.search

object BinarySearch {
    fun isValueInArray(arrayInt: IntArray, value: Int): Int {
        var left = 0
        var right = arrayInt.size - 1
        while (left <= right) {
            val middle = (left + right) / 2
            val current = arrayInt[middle]
            if (current == value) return middle
            else if (current < value) left = middle + 1
            else right = middle - 1
        }
        return -1
    }
}